### ISSUE TITLE
Bug: False positive `markdown/no-missing-label-refs` on GFM Alerts

### ISSUE THREAD

**Issue by midskyey** [eslint/markdown#294]
### Environment

ESLint version: 9.12.0
@eslint/markdown version: 6.2.1
Node version: 22.9.0
npm version: 10.8.3
Operating System: macOS 14.7

### Which language are you using?

gfm

### What did you do?

<details>
<summary>Configuration</summary>

```js
import pluginMd from "@eslint/markdown"

export default [
  // Silly type narrowing for spreading in full-fledged typechecked `eslint.config.js`
  ...[pluginMd.configs?.recommended].filter(Boolean).flat(),
  {
    files: ["**/*.md"],
    language: "markdown/gfm",
  },
]

```
</details>

```md
> [!NOTE]
> Useful information that users should know, even when skimming content.

> [!TIP]
> Helpful advice for doing things better or more easily.

> [!IMPORTANT]
> Key information users need to know to achieve their goal.

> [!WARNING]
> Urgent info that needs immediate user attention to avoid problems.

> [!CAUTION]
> Advises about risks or negative outcomes of certain actions.
```

Which is being rendered on GH as:

> [!NOTE]
> Useful information that users should know, even when skimming content.

> [!TIP]
> Helpful advice for doing things better or more easily.

> [!IMPORTANT]
> Key information users need to know to achieve their goal.

> [!WARNING]
> Urgent info that needs immediate user attention to avoid problems.

> [!CAUTION]
> Advises about risks or negative outcomes of certain actions.

### What did you expect to happen?

No violations.

### What actually happened?

```
   1:4  error  Label reference '!NOTE' not found       markdown/no-missing-label-refs
   4:4  error  Label reference '!TIP' not found        markdown/no-missing-label-refs
   7:4  error  Label reference '!IMPORTANT' not found  markdown/no-missing-label-refs
  10:4  error  Label reference '!WARNING' not found    markdown/no-missing-label-refs
  13:4  error  Label reference '!CAUTION' not found    markdown/no-missing-label-refs
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/118f65b5-5b7b-4190-9fbe-b023b0542a4e?file=package.json

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

More info on GFM Alerts:
- https://github.com/orgs/community/discussions/16925
- https://github.blog/changelog/2023-12-14-new-markdown-extension-alerts-provide-distinctive-styling-for-significant-content/
- https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

**Comment by mdjermanovic** [comment]
Thanks for the issue! I can reproduce this with `@eslint/markdown` v6.2.1 (and v6.2.0 as well, so this isn't a regression introduced in v6.2.1).

I think this should be fixed in the parser as it treats `[!NOTE]` and others as just plain text, while it should probably be a new node type. @eslint/eslint-team thoughts?

**Comment by nzakas** [comment]
Yes, this is an issue with the parser not supporting GitHub alerts. It needs to be fixed there.  Opened: https://github.com/syntax-tree/mdast-util-gfm/issues/2

**Comment by nzakas** [comment]
So it looks like alerts are not actually part of GFM and are just an HTML transform:
https://github.com/syntax-tree/mdast-util-gfm/issues/2#issuecomment-2422827671

By "HTML transform", they mean that the parsing of the Markdown isn't any different, it's just the way that the Markdown is interpreted while transforming to HTML that changes.

Here's the GFM spec:
https://github.github.com/gfm/

We have a couple options for what to do here:

1. We could create a plugin that generates a separate AST for this syntax to make it more easily identifiable in the AST.
2. We could add an option to the rule that filters out known GitHub alert labels.

**Comment by JoshuaKGoldberg** [comment]
Agreed on 1, this feels like a plugin to me. GitHub alert labels are just one example of a nonstandard extension to GFM. Folks writing `.md` for other platforms won't want them.

**Comment by mdjermanovic** [comment]
> 1. We could create a plugin that generates a separate AST for this syntax to make it more easily identifiable in the AST.

We would then always use this plugin when [parsing in `gfm` mode](https://github.com/eslint/markdown/blob/05196f59dd1d22e7bcc98ee5e5b857e69807a5ce/src/language/markdown-language.js#L108-L112)? That seems like a nice solution, though our parser would be doing something unique because other markdown parsers treat this as plain text, so that might be surprising to rule authors?

**Comment by nzakas** [comment]
> We would then always use this plugin when [parsing in `gfm` mode](https://github.com/eslint/markdown/blob/05196f59dd1d22e7bcc98ee5e5b857e69807a5ce/src/language/markdown-language.js#L108-L112)? That seems like a nice solution, though our parser would be doing something unique because other markdown parsers treat this as plain text, so that might be surprising to rule authors?

Yes. I think we should maintain the sanctity of what "GFM" means and ensure it matches the spec by default. If we want to create an AST representation of alerts, then I think that should be opt-in as a language option for `markdown/gfm`.

**Comment by lumirlumir** [comment]
I've opened a discussion in the syntax-tree organization: https://github.com/orgs/syntax-tree/discussions/144

If this feature request is accepted, the issue would be easy to address.

**Comment by lumirlumir** [comment]
@nzakas @mdjermanovic  

I believe this bug is not related to the GFM specification.  

The issue with the `no-missing-label-refs` rule is that it does not internally use the `LinkReference` and `ImageReference` node types.  

```js  
/*  
 * Matches substrings like "[foo]", "[]", "[foo][bar]", "[foo][]", "[][bar]", or "[][]".  
 * `left` is the content between the first brackets. It can be empty.  
 * `right` is the content between the second brackets. It can be empty, and it can be undefined.  
 */  
const labelPattern = /\[(?<left>[^\]]*)\](?:\[(?<right>[^\]]*)\])?/dgu;  
```  

The current implementation attempts to parse the `Text` node directly. However, this can lead to bugs when using GitHub Alerts syntax, as `> [!TIP]` is mistakenly recognized as a `labelPattern`.  

To fix this, we should use the `LinkReference` and `ImageReference` node types to find the correct nodes instead of relying on regular expressions to parse `Text` nodes.

**Comment by lumirlumir** [comment]
@nzakas Ah, I missed it. Thanks for the example.

**Comment by lumirlumir** [comment]
Hello, @samanthacomes 

Front Matter Parsing option has been added in v6.4.0. (https://github.com/eslint/markdown/pull/328)

Now, you can set `languageOptions.frontmatter: 'yaml'` to correctly parse YAML front matter and avoid false positive.

**Comment by lumirlumir** [comment]
@nzakas I've been thinking about the correct solution for a few days, and I believe the best approach to this problem is to provide users with a way to ignore specific labels.

For example:

```js
'markdown/no-missing-label-refs': ["error", { ignore: ['!NOTE', '!TIP', '!IMPORTANT'] }]
```

With this, we can add support for ignoring specific labels.

If this suggestion sounds good, I can open a PR within a week!

**Comment by nzakas** [comment]
@lumirlumir thanks for the suggestion. I'm not sure that makes sense because block quotes don't really have defined labels in markdown (spec: https://spec.commonmark.org/0.31.2/#block-quotes). We'd just be overlaying a GitHub-specific meaning on top of this, which I don't think is the right approach. I also think people expect that in GFM this type of behavior would be automatic, rather than requiring all users to add this configuration.

**Comment by lumirlumir** [comment]
I'm back to revisit this issue. Since this bug currently requires certain files in `@eslint/markdown` to include `eslint-disable` directives, I think it would be nicer to decide how we want to proceed with this issue.

* https://github.com/eslint/markdown/blob/76685c58eab498bf4b4b2661d9fb303f37f18303/docs/rules/no-duplicate-definitions.md?plain=1#L18
* https://github.com/eslint/markdown/blob/76685c58eab498bf4b4b2661d9fb303f37f18303/docs/rules/no-bare-urls.md?plain=1#L16
* https://github.com/eslint/markdown/blob/76685c58eab498bf4b4b2661d9fb303f37f18303/docs/rules/table-column-count.md?plain=1#L11

Since https://github.com/eslint/rewrite/issues/221 hasn’t progressed, I considered some alternatives — but ultimately, I think there may be only one viable solution.

### ✅ Solution: Add an `allowLabels` option (as suggested in [the comment above](https://github.com/eslint/markdown/issues/294#issuecomment-2835608666))

The `no-missing-label-refs` rule is based on Markdownlint’s [MD052](https://github.com/DavidAnson/markdownlint/blob/main/doc/md052.md). That rule includes an `ignored_labels` option, which behaves similarly to what was suggested in the earlier comment.

![Image](https://github.com/user-attachments/assets/918a5337-f3b3-44aa-8270-0e062894363a)

This approach would be intuitive for users and easier to implement than the alternatives.

Since many of the rules in `@eslint/markdown` follow the implementations of Markdownlint and Remarklint, I think it's a good approach that allows users to align their rules with existing Markdown linters.

### ⚠️ Alternative: Add a custom MDAST plugin that understands GitHub Alert Syntax

I previously shared a [discussion](https://github.com/orgs/syntax-tree/discussions/144) about this, but the consensus was that this syntax is a purely HTML-based transform.

Even when reviewing the [GFM spec](https://github.github.com/gfm/), GitHub Alert blocks are not officially recognized, since they are not part of the specification.

Although GitHub supports this custom Alert Syntax, it hasn’t been incorporated into the GFM spec (which hasn’t been updated since 2019 😅). Supporting it via a custom plugin could introduce inconsistencies, as it would be deviating from standardized behavior.

---

So in conclusion, I'm proposing to add an `allowLabels` option, and I'd like to hear other team members' thoughts on this approach — @eslint/eslint-team.

Once we reach a consensus on how to proceed with this issue, I'd be happy to take a look at it.

**Comment by JoshuaKGoldberg** [comment]
I don't think the ideal solution here is to add anything _rule_-specific. `markdown/no-missing-label-refs` might be the only rule impacted by this issue now, but it's likely there'll be other rules later on that interact with these kinds of nodes. I think the right solution would be to add a way to let users opt into the alternate parsing mode. Perhaps a `parserOptions` option?

It's kind of unfortunate that the current implementation in userland for the GitHub alerts is an html transform, https://github.com/orgs/syntax-tree/discussions/144#discussioncomment-12652504 -> [rehype-github-alert](https://github.com/rehypejs/rehype-github/tree/main/packages/alert). But:

> Although GitHub supports this custom Alert Syntax, it hasn’t been incorporated into the GFM spec (which hasn’t been updated since 2019 😅). Supporting it via a custom plugin could introduce inconsistencies, as it would be deviating from standardized behavior.

GitHub doesn't change their Markdown parsing very often. And as noted, the spec changes even less often. This is relatively pretty small addition to the spec (compared to, say, the TypeScript types work in core). I think writing that custom plugin would be the right way to go.

If the custom plugin is just absolutely not doable, I suppose the `allowLabels` option would be a reasonable fallback. But it's kind of odd to add an option only for one use case that has a different "ideal" solution.

**Comment by lumirlumir** [comment]
Ah, I truly understood what the team aims for 👍 Thank you all for the detailed comments.

Then, maybe creating the spec for how to treat GitHub Alert syntax, and apply it by creating custom `mdast-util-gfm-alert` or similar would make sense?

**Comment by nzakas** [comment]
Yes, it would be great for someone to propose something. :smile: We probably don't want to get into having a separate package unless we get a request for it.

**Comment by lumirlumir** [comment]
Sounds good to me 👍 If it's okay, I’d be happy to take a look at this.

**Comment by nzakas** [comment]
Sounds good. Just to be clear: we'd still like you to describe an idea on this issue *before* submitting a PR to make sure everyone is aligned on the approach.

**Comment by lumirlumir** [comment]
@nzakas I'll make sure to create an RFC document before proceeding, just like we did in https://github.com/eslint/markdown/pull/230.

**Comment by lumirlumir** [comment]
RFC has been opened: https://github.com/eslint/rfcs/pull/138
